print_nanny_webapp/alerts/consumers.py

```diff
@@ -29,3 +41,57 @@ def alert_message(self, message):
         self.send(message["data"])
 
         logger.info(f"Received message {message}")
+
+
+class DiscordConsumer(AsyncConsumer):
+    groups = []
+
+    async def trigger_alert(self, data):
+        logger.info(f"Received message for user IDs {data['user_ids']} and channel IDs {data['channel_ids']}")
+        await discord_client.wait_until_ready()
+
+        for user in data["user_ids"]:
+            target = await discord_client.fetch_user(int(user))
+            # TODO: Return an error
+            await self._target_send(user, target, data["message"])
+
+        for channel in data["channel_ids"]:
+            target = await discord_client.fetch_channel(int(channel))
+            await self._target_send(channel, target, data["message"])
+
+    async def _target_send(self, _id, target, message):
+            if target is None:
+                # TODO: Return an error
+                logger.error(f"Discord could not find item '{_id}'!")
+                return None
+
+            return await target.send(message)
+
+    async def validate_id(self, data):
+        logger.info(f"Validating {data['target_id_type']} ID {data['target_id']}")
+        await discord_client.wait_until_ready()
+        res = {
+            "is_valid": True,
+            "value": data["target_id"],
+            "error": ""
+        }
+
+        fetcher = None
+        if data["target_id_type"] == "USER":
+            fetcher = discord_client.fetch_user
+        elif data["target_id_type"] == "CHANNEL":
+            fetcher = discord_client.fetch_channel
+        else:
+            res["is_valid"] = False
+            res["error"] = "Unknown ID type"
+            await self.base_send(res)
+
+
+        try:
+            await fetcher(int(data["target_id"]))
+        except (InvalidData, HTTPException, NotFound, Forbidden) as e:
+            res["is_valid"] = False
+            res["error"] = f"During '{data['target_id']}' {data['target_id_type']} validation, the following error occurred: {e}"
+            await self.base_send(res)
+
+        await self.base_send(res)
```

Comment:
@leigh-johnson I'm unable to get this to work. It reports:
```python
ERROR 2021-02-11 17:59:31,237 server 1 140618291976000 Exception inside application: You must implement application_send()
  File "/usr/local/lib/python3.8/site-packages/channels/routing.py", line 71, in __call__
    return await application(scope, receive, send)
  File "/usr/local/lib/python3.8/site-packages/channels/routing.py", line 190, in __call__
    return await application(scope, receive, send)
  File "/usr/local/lib/python3.8/site-packages/channels/consumer.py", line 94, in app
    return await consumer(scope, receive, send)
  File "/usr/local/lib/python3.8/site-packages/channels/consumer.py", line 58, in __call__
    await await_many_dispatch(
  File "/usr/local/lib/python3.8/site-packages/channels/utils.py", line 51, in await_many_dispatch
    await dispatch(result)
  File "/usr/local/lib/python3.8/site-packages/channels/consumer.py", line 73, in dispatch
    await handler(message)
  File "/app/print_nanny_webapp/alerts/consumers.py", line 95, in validate_id
    await self.base_send(res)
  File "/usr/local/lib/python3.8/site-packages/asgiref/server.py", line 73, in application_send
    raise NotImplementedError("You must implement application_send()")
  You must implement application_send()
```

I can't seem to fine any way of responding to a message with the `AsyncConsumer` (while the `WebsocketAsyncConsumer` does exactly the same under the hood...). Ideas?

Reply:
Which line of code calls `validate_ids`?